text/0000-nano-contracts.md

```diff
@@ -0,0 +1,431 @@
+│                     │  make_a_bet(BRA)                     │
+├─────────────────────┼──────────────────────────────────────┤
+│                     │  500 HTR                             │
+│                     │  P2PKH                               │
+│                     │  HDEJ453QjK7F7jipQebRymsRpnFG8j2DHk  │
+└─────────────────────┴──────────────────────────────────────┘
+</pre>
+
+To clarify, transactions will only be valid if the sum of the inputs equals the sum of the outputs.
+
+### Validation Time
+
+Transactions are validated as soon as they arrive in the full node. If they are valid, they are immediately added to the DAG of transactions. Otherwise, they are discarded.
+
+On the other hand, as to nano contracts, the result of a call to a method depends on the order the calls are executed. It means that the full node can only execute calls when the order of execution is well defined and the same for all peers in the network.
+
+The order of execution cannot depend on the order the calls arrived in the full node. So, when can a call be safely executed? Only after it is confirmed by a block. It means that calls in the mempool must wait to be confirmed.
+
+### Order of Execution
+
+The calls execution order is defined by the accumulated weight of the transactions, from higher to lower.
+
+Inside a transaction, the calls are executed following the order of the inputs and then following the order of the outputs.
+
+For instance, let's say all transactions in the image below are making calls to nano contracts. The order of execution will be:
+
+![fig-nano-contract-execution-order](./0000-images/nano-contracts-execution-order.svg "Nano Contract Execution Order")
+
+TODO Add `acc_weight` to each tx in the image above and explain the ordering here.
+
+## Examples of Use
+
+### Group Betting
+
+Betting using data from the real world requires an Oracle. Imagine a predict-a-score game where any person can place a bet and the funds will be split among the winners. In this case, each participants will transfer funds to the Nano Contract. At the end of the game, an Oracle will provide the game results and the winners will be able to collect their prize. Such Nano Contract would have the following blueprint:
+
+```python
+class BetItem:
+	total: Amount = 0
+	bets: Dict[Address, Amount] = 0
+
+	def add_bet(self, address: Address, amount: Amount):
+		self.bets[address] += amount
+		self.total += amount
+
+class Bet(Blueprint):
+	bets: Dict[str, BetItem]
+	withdrawals: Dict[Address, Amount]
+	total: Amount
+	oracle_script: TxOutputScript
+	date_last_offer: Timestamp
+	result: str
+
+	@public
+	def make_a_bet(self, address: Address, amount: Amount, score: str) -> None:
+		"""Make a bet."""
+		if timestamp > date_last_offer:
+			self.fail('cannot place bets after 2021-01-01 17:00')
+		self.bets[score].add_bet(address, amount)
+		self.total += amount
+
+	@public
+	def set_result(self, result: str) -> None:
+		"""Set result. Used by the oracle."""
```

Comment:
Only the oracle can call this? Would be interesting if anyone can call it, as long as the message is properly signed by the oracle.